libraries/AP_OpenDroneID/AP_OpenDroneID_DroneCAN.cpp

```diff
@@ -0,0 +1,279 @@
+
+    dc_system[driver_index] = new uavcan::Publisher<dronecan::remoteid::System>(*uavcan->get_node());
+    if (dc_system[driver_index] == nullptr) {
+        goto alloc_failed;
+    }
+    dc_system[driver_index]->setTxTimeout(uavcan::MonotonicDuration::fromMSec(20));
+    dc_system[driver_index]->setPriority(uavcan::TransferPriority::OneHigherThanLowest);
+
+    dc_operator_id[driver_index] = new uavcan::Publisher<dronecan::remoteid::OperatorID>(*uavcan->get_node());
+    if (dc_operator_id[driver_index] == nullptr) {
+        goto alloc_failed;
+    }
+    dc_operator_id[driver_index]->setTxTimeout(uavcan::MonotonicDuration::fromMSec(20));
+    dc_operator_id[driver_index]->setPriority(uavcan::TransferPriority::OneHigherThanLowest);
+
+    arm_status_listener[driver_index] = new uavcan::Subscriber<dronecan::remoteid::ArmStatus, ArmStatusCb>(*uavcan->get_node());
+    if (arm_status_listener[driver_index] == nullptr) {
+        goto alloc_failed;
+    }
+    arm_status_listener[driver_index]->start(ArmStatusCb(uavcan, &handle_arm_status));
+
+    dronecan_done_init |= driver_mask;
+    return;
+
+alloc_failed:
+    dronecan_init_failed |= driver_mask;
+    GCS_SEND_TEXT(MAV_SEVERITY_NOTICE, "OpenDroneID DroneCAN alloc failed");
+}
+
+/*
+  send pending DroneCAN OpenDroneID packets
+ */
+void AP_OpenDroneID::dronecan_send(AP_UAVCAN *uavcan)
+{
+    const uint8_t driver_index = uavcan->get_driver_index();
+    const uint8_t driver_mask = 1U<<driver_index;
+
+    if (driver_index+1 != _can_driver) {
+        // not enabled for this CAN driver
+        return;
+    }
+
+    dronecan_init(uavcan);
+
+    if (dronecan_init_failed & driver_mask) {
+        return;
+    }
+
+    if (need_send_basic_id & driver_mask) {
+        WITH_SEMAPHORE(_sem);
+        dronecan_send_basic_id(uavcan);
+        need_send_basic_id &= ~driver_mask;
+    }
+    if (need_send_system & driver_mask) {
+        WITH_SEMAPHORE(_sem);
+        dronecan_send_system(uavcan);
+        need_send_system &= ~driver_mask;
+    }
+    if (need_send_self_id & driver_mask) {
+        WITH_SEMAPHORE(_sem);
+        dronecan_send_self_id(uavcan);
+        need_send_self_id &= ~driver_mask;
+    }
+    if (need_send_operator_id & driver_mask) {
+        WITH_SEMAPHORE(_sem);
+        dronecan_send_operator_id(uavcan);
+        need_send_operator_id &= ~driver_mask;
+    }
+    if (need_send_location & driver_mask) {
+        WITH_SEMAPHORE(_sem);
+        dronecan_send_location(uavcan);
+        need_send_location &= ~driver_mask;
+    }
+}
+
+#define ODID_COPY(name) msg.name = pkt.name
+#define ODID_COPY_STR(name) do { for (uint8_t i = 0; i<sizeof(pkt.name) && pkt.name[i]; i++) msg.name.push_back(pkt.name[i]); } while(0)
+
+
+void AP_OpenDroneID::dronecan_send_location(AP_UAVCAN *uavcan)
+{
+    dronecan::remoteid::Location msg {};
+    const auto &pkt = pkt_location;
+    ODID_COPY_STR(id_or_mac);
+    ODID_COPY(status);
```

Comment:
@rabad-matternet @tridge , just checking, will AP set this field automatically in the event of emergency status?

Where Emergency Status is:
UAS unable to recover from an uncontrolled descent
UAS unable to recover from loss of control of the flight trajectory

Is this triggered by an FTS? Can we modify this field manually for testing?